library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -2793,8 +2793,12 @@ impl<T> From<Vec<T>> for VecDeque<T> {
             let len = other.len();
 
             // We need to extend the buf if it's not a power of two, too small
-            // or doesn't have at least one free space
-            if !buf.capacity().is_power_of_two()
+            // or doesn't have at least one free space.
+            // We check if `T` is a ZST in the first condition,
+            // because `usize::MAX` (the capacity returned by `capacity()` for ZST)
+            // is not a power of two and thus it'll always try
+            // to reserve more memory which will panic for ZST (rust-lang/rust#78532)
+            if (!buf.capacity().is_power_of_two() && mem::size_of::<T>() != 0)
```

Comment:
Maximum len of ZST Vec is larger than maximum len of ZST VecDeque (MAXIMUM_ZST_CAPACITY - 1), which would be another thing to check before constructing VecDeque.

Reply:
I'm not sure if this should be changed in this PR, but it's definitely a check that should be added I think.

Reply:
Fair. I mentioned it here, because it wasn't an issue so far, but will be as a result of changes in this PR.

Reply:
You can just open a new issue👍

Reply:
We now create a `VecDeque` with a capacity which is not a power of two here, breaking one of its invariants.

This is either unsound or that invariant isn't actually required anywhere and we should remove it.